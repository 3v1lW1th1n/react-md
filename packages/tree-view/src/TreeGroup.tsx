import * as React from "react";
import { Collapse } from "@react-md/transition";
import { List, IListProps, ListElement } from "@react-md/list";

export interface ITreeGroupProps extends IListProps {
  /**
   * An optional configurable enter duration. This defaults to the collapse's enter duration
   * of `250ms`.
   *
   * @docgen
   */
  enterDuration?: number;

  /**
   * An optional configurable leave duration. This defaults to the collapse's leave duration
   * of `200ms`.
   *
   * @docgen
   */
  leaveDuration?: number;

  /**
   * Boolean if the group should be removed from the DOM when the group is not expanded.
   *
   * @docgen
   */
  isEmptyCollapsed?: boolean;

  /**
   * Boolean if the group is currently expanded.
   *
   * @docgen
   */
  expanded: boolean;
}

export interface ITreeGroupDefaultProps {
  isEmptyCollapsed: boolean;
}

export type TreeGroupWithDefaultProps = ITreeGroupProps & ITreeGroupDefaultProps;

const TreeGroup: React.SFC<ITreeGroupProps> = providedProps => {
  const {
    style,
    className,
    expanded,
    children,
    isEmptyCollapsed,
    enterDuration,
    leaveDuration,
    ...props
  } = providedProps as TreeGroupWithDefaultProps;

  return (
    <Collapse
      style={style}
      className={className}
      collapsed={!expanded}
      isEmptyCollapsed={isEmptyCollapsed}
      enterDuration={enterDuration}
      leaveDuration={leaveDuration}
    >
      {({ refCallback, ...collapseProps }) => (
        <List {...props} role="group" {...collapseProps} ref={refCallback}>
          {children}
        </List>
      )}
    </Collapse>
  );
};

TreeGroup.defaultProps = {
  isEmptyCollapsed: true,
} as ITreeGroupDefaultProps;

export default TreeGroup;
