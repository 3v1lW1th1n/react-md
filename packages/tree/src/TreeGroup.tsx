import React, { FunctionComponent } from "react";
import cn from "classnames";
import { Collapse, CollapseProps } from "@react-md/transition";
import { List, ListProps } from "@react-md/list";
import { Omit } from "@react-md/utils";

export interface TreeGroupProps
  extends ListProps,
    Omit<CollapseProps, "children"> {}

/**
 * The `TreeGroup` component is used to render a tree item's nested items whenever the `expanded`
 * prop is `true`. It uses the `Collapse` component behind the scenes to animate in-and-out of view
 * and will fully unrender when the `expanded` prop is `false`.
 */
const TreeGroup: FunctionComponent<TreeGroupProps> = ({
  style,
  className,
  collapsed,
  minHeight,
  minPaddingBottom,
  minPaddingTop,
  enterDuration,
  leaveDuration,
  isEmptyCollapsed,
  onExpanded,
  onCollapsed,
  children,
  ...props
}) => (
  <Collapse
    style={style}
    className={cn("rmd-tree-group", className)}
    collapsed={collapsed}
    minHeight={minHeight}
    minPaddingBottom={minPaddingBottom}
    minPaddingTop={minPaddingTop}
    enterDuration={enterDuration}
    leaveDuration={leaveDuration}
    isEmptyCollapsed={isEmptyCollapsed}
    onExpanded={onExpanded}
    onCollapsed={onCollapsed}
  >
    <List {...props} role="group">
      {children}
    </List>
  </Collapse>
);

TreeGroup.defaultProps = {
  isEmptyCollapsed: true,
};

export default TreeGroup;
