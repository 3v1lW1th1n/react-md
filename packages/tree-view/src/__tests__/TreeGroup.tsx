import * as React from "react";
import renderer from "react-test-renderer";

import TreeGroup from "../TreeGroup";

// TODO: Write more tests once enzyme supports the new react context API.
describe("TreeGroup", () => {
  it("should render correctly with the provided props", () => {
    expect(
      renderer
        .create(
          <TreeGroup expanded={false}>
            <li>Something!</li>
          </TreeGroup>
        )
        .toJSON(`null`)
    ).toMatchSnapshot();
    expect(
      renderer
        .create(
          <TreeGroup expanded={true}>
            <li>Something!</li>
          </TreeGroup>
        )
        .toJSON(`null`)
    ).toMatchSnapshot();
  });
});
