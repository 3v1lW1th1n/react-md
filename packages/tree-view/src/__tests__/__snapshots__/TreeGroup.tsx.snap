// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeGroup should render correctly with the provided props 1`] = `null`;

exports[`TreeGroup should render correctly with the provided props 2`] = `
<ul
  className="rmd-list rmd-collapse"
  role="group"
>
  <li>
    Something!
  </li>
</ul>
`;
